simphony/cuds/store.py

```diff
@@ -0,0 +1,89 @@
+"""Base storage mechanism for accessing CUDS entities, mainly datasets.
+
+This module contains the base class for implementing different access
+strategies for CUDS,notably MemoryStore and ProxyStore.
+"""
+import abc
+
+
+class ABCStateDataStore(object):
+    """Abstract base class for all stores.
+    """
+    __metaclass__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def add(self, dataset, *args, **kwargs):
+        """Add the dataset to the store."""
+
+    @abc.abstractmethod
+    def get(self, name):
+        """Return the dataset with the given name."""
+
+    @abc.abstractmethod
+    def remove(self, name):
+        """Remove the dataset with the given name."""
+
+    @abc.abstractmethod
+    def get_names(self):
+        """Return names of the existing datasets."""
+
+    @abc.abstractmethod
+    def iter_datasets(self, names=None):
+        """Iterate over all or a subset of datasets."""
+
+
+class MemoryStateDataStore(ABCStateDataStore):
+    """A dataset store that keeps everything in memory."""
+    def __init__(self, *args, **kwargs):
+        self._datasets = {}
+
+    def add(self, dataset, *args, **kwargs):
+        """Add the dataset to the store."""
+        self._datasets[dataset.name] = dataset
+
+    def get(self, name):
+        """Return the dataset with the given name."""
+        return self._datasets.get(name)
+
+    def remove(self, name):
+        """Remove the dataset with the given name."""
+        del self._datasets[name]
+
+    def get_names(self):
+        """Return names of the existing datasets."""
+        return self._datasets.keys()
+
+    def iter_datasets(self, names=None):
+        """Iterate over all or a subset of datasets."""
+        if not names:
+            names = self._datasets.keys()
+
+        for key in self._datasets:
+            if key in names:
+                yield self._datasets.get(key)
+
```

Comment:
Since `key` is already read from the `_dataset` dictionary itself, it will always have a value and no KeyError will raise. However I changed it, in order to stay on the safer side :)

For the `get_method` method, it is intentional, similar to python's `dict.get` method.